website/source/api/features/filtering.html.md

```diff
@@ -0,0 +1,458 @@
+base 10 integers and floating point numbers are possible. For quoted strings,
+they may either be enclosed in double quotes or backticks. When enclosed in
+backticks they are treated as raw strings and escape sequences such as `\n`
+will not be expanded.
+
+## Filter Utilization
+
+Generally, only the main object is filtered. When filtering for
+an item within an array, that is not at the top level, the entire array that contains the item
+will be returned. This is usually the outermost object of a response,
+but in some cases such the [`/catalog/node/:node`](api/catalog.html#list-services-for-node)
+endpoint the filtering is performed on a object embedded within the results.
+
+### Performance
+
+Filters are executed on the servers and therefore will consume some amount
+of CPU time on the server. For non-stale queries this means that the filter
+is executed on the leader.
+
+### Filtering Examples
+
+#### Agent API
+
+**Command - Unfiltered**
+
+```sh
+curl -X GET localhost:8500/v1/agent/services
+```
+
+**Response - Unfiltered**
+
+```json
+{
+    "redis1": {
+        "ID": "redis1",
+        "Service": "redis",
+        "Tags": [
+            "primary",
+            "production"
+        ],
+        "Meta": {
+            "env": "production",
+            "foo": "bar"
+        },
+        "Port": 1234,
+        "Address": "",
+        "Weights": {
+            "Passing": 1,
+            "Warning": 1
+        },
+        "EnableTagOverride": false
+    },
+    "redis2": {
+        "ID": "redis2",
+        "Service": "redis",
+        "Tags": [
+            "secondary",
+            "production"
+        ],
+        "Meta": {
+            "env": "production",
+            "foo": "bar"
+        },
+        "Port": 1235,
+        "Address": "",
+        "Weights": {
+            "Passing": 1,
+            "Warning": 1
+        },
+        "EnableTagOverride": false
+    },
+    "redis3": {
+        "ID": "redis3",
+        "Service": "redis",
+        "Tags": [
+            "primary",
+            "qa"
+        ],
+        "Meta": {
+            "env": "qa"
+        },
+        "Port": 1234,
+        "Address": "",
+        "Weights": {
+            "Passing": 1,
+            "Warning": 1
+        },
+        "EnableTagOverride": false
+    }
+}
+```
+
+**Command - Filtered**
+
+```sh
+curl -G localhost:8500/v1/agent/services --data-urlencode 'filter=Meta.env == qa'
+```
+
+**Response - Filtered**
+
+```json
+{
+    "redis3": {
+        "ID": "redis3",
+        "Service": "redis",
+        "Tags": [
+            "primary",
+            "qa"
+        ],
+        "Meta": {
+            "env": "qa"
+        },
+        "Port": 1234,
+        "Address": "",
+        "Weights": {
+            "Passing": 1,
+            "Warning": 1
+        },
+        "EnableTagOverride": false
+    }
+}
+```
+
+#### Catalog API
+
+**Command - Unfiltered**
+
+```sh
+curl -X GET localhost:8500/v1/catalog/service/api-internal
```

Comment:
Maybe it would be clearer to settle on either `-G` or `-X GET` but not a random mix of both in the examples provided here.

Reply:
`-G` and `-X GET` are not functionally equivalent.

From the man pages:

For `-G`
```
              When used, this option will make all data specified with -d, --data, --data-binary or
              --data-urlencode  to  be used in an HTTP GET request instead of the POST request that
              otherwise would be used. The data will be appended to the URL with a '?' separator.
```

For `-X`
```
              This  option only changes the actual word used in the HTTP request, it does not alter
              the way curl behaves. So for example if you want to make a proper HEAD request, using
              -X HEAD will not suffice. You need to use the -I, --head option.
```

In general we could omit the `-X GET` as it will do a GET by default. However after talking with @kaitlincarter-hc the consensus was that to be as consistent with the rest of our docs we should have it in there. For all the queries where
`--data-urlencode` is used, curl would by default make it a POST request and put the data in the body of the request. Using `-X GET` makes it a GET request but the filter is then still transmitted in the body. Using `-G` forces curl to shove the filter in as a query param while still performing the urlencoding.

This seems like a deficiency with curl not being able to urlencode query params easier but its all we have to work with right now.

Also of note is that in all the cases where we are using `-X GET` we could use `-G` so we could standardize on that here. @kaitlincarter-hc what do you think?